librubyfmt/src/lib.rs

```diff
@@ -76,28 +76,30 @@ pub enum FormatError {
     OtherRubyError = 4,
 }
 
+// FIXME: Why does this need to be a string? We're just printing it to stdout
 pub fn format_buffer(buf: &str) -> Result<String, RichFormatError> {
-    let tree = run_parser_on(buf)?;
+    let (tree, file_comments) = run_parser_on(buf)?;
     let out_data = vec![];
     let mut output = Cursor::new(out_data);
-    let data = buf.as_bytes();
-    toplevel_format_program(&mut output, data, tree)?;
-    output.flush().expect("flushing works");
-    Ok(unsafe { String::from_utf8_unchecked(output.into_inner()) })
+    toplevel_format_program(&mut output, tree, file_comments)?;
+    output.flush().expect("flushing to a vec should never fail");
+    Ok(String::from_utf8(output.into_inner()).expect("we never write invalid UTF-8"))
 }
 
 #[no_mangle]
 pub extern "C" fn rubyfmt_init() -> libc::c_int {
     init_logger();
-    unsafe {
-        ruby::ruby_init();
+    let res = ruby_ops::setup_ruby();
+    if res.is_err() {
+        return InitStatus::ERROR as libc::c_int;
```

Comment:
Should the C API provide a way to get a human readable description of the error?

```rust
static mut LAST_ERROR: Option<CString> = None;

// on this line
if let Err(e) = res {
    // If we think there could be multi-threading here we should use `lazy_static!` and a mutex
    unsafe { LAST_ERROR = Some(CString::new(e.to_string()).expect("don't actually expect in prod")) }
    return InitStatus::ERROR as libc::c_int;
}

// later
#[no_mangle]
pub extern "C" fn rubyfmt_last_error -> *const libc::c_char {
    match LAST_ERROR {
        Some(e) => e.as_ptr(),
        None => ptr::null(),
    }
}
```

Reply:
the header defines this: https://github.com/penelopezone/rubyfmt/blob/trunk/librubyfmt/include/rubyfmt.h#L4